### PR TITLE
Doctrine migrations version 3

### LINKED ISSUE #25: Upgrade to doctrine-migrations v3
Given the release of [Doctrine Migrations v3](https://github.com/doctrine/migrations/releases/tag/3.0.0), is the support for the new version planned? It's such a pity, because the migrations support for `psr-container-doctrine` was added not a long time ago, but the version is now locked to v2. Can we do this without a BC break?

### PR DESCRIPTION
**Pull request by edigu**
[doctrine/migrations v3](https://github.com/doctrine/migrations/releases/tag/3.0.0) released a while ago and it introduces backward-incompatible changes with 2.x release line.
This draft aims adapting container factories introduced recently for migrations v3 to close #25 

Summary of notable changes are:
 - BC BREAK! This changes will require a major version upgrade to 3.x and it is not possible to support both migration versions in same release line.
 - Add two new factories
   -  `DependencyFactoryFactory` - responsible for creating [DependencyFactory](https://github.com/doctrine/migrations/blob/3.0.x/lib/Doctrine/Migrations/DependencyFactory.php) instances
   - `ConfigrationLoaderFactory` - creates [ConfigurationArray](https://github.com/doctrine/migrations/blob/3.0.x/lib/Doctrine/Migrations/Configuration/Migration/ConfigurationArray.php) instances based on the array configration provided.
 - Bumped PHPUnit version (from 8.4 > 9.4) since minimum PHP version required in composer.json was already `php >= 7.3`
 - `ConnectionFactory` was relying on deprecated DBAL namespaces. Changed imports with suggested namespaces instead.
 - All migration related factories moved to `Roave\PsrContainerDoctrine\Migrations` sub-namespace
 - Adapted docs
 - Added / adapted tests, migrated `phpunit.xml.dist` to 9.4 format

I choose master as merge target for now since there were no 3.x branch available.

### CODE REVIEW

**Comment by Moln** [comment]
`composer.json` need to upgrade : 

```
"doctrine/common": "^2.6",
```

to

```
"doctrine/common": "^2.6 || ^3.0",
```

File: composer.json
Comment (Moln) [inline composer.json:20]:
```
"doctrine/common": "^2.6 || ^3.0",
```

Comment (edigu) [reply composer.json:20]: Umm this is intersting. Why? Does it also applies to `dbal`? I mean for the consistency, do we need to provide more relaxed library versions? If the answer is yes, what would be other good candidates?

Comment (Ocramius) [reply composer.json:20]: I don't mind bumping dependency ranges and restricting.

A good strategy to ease upgrading is relaxing in one minor, then bumping in another minor

Comment (Moln) [reply composer.json:20]: Sorry, it's my fault. I thought it was a PR for upgrade to ORM `2.8`.

**Comment by PowerKiKi** [comment]
@edigu this PR is marked as draft. Did you have something else to do before a formal reviewing ?

**Comment by edigu** [comment]
> @edigu this PR is marked as draft. Did you have something else to do before a formal reviewing ?

@PowerKiKi thank you for reminding, formally it's ready for review from implementation point of view. There are still some work left to satisfy psalm which stressing me (I would appreciate some help there). I was busy as hell with daily work recently, just changed the status.

I am also considering relaxing dependencies as we discussed above for `doctrine/common` and `dbal`

**Comment by edigu** [comment]
Okay I relaxed constraints for `doctrine/common` and `dbal` in latest state to allow more recent versions when its needed. Tests are green, however both psalm and I are unhappy with remaining analysis warning. I need some guidance to address 4 remaining warnings or get a green light to introduce more `pslam-suppress` annotations which makes me feel bad.

**Comment by ceadreak** [comment]
Hello ! I'm waiting for this feature too to upgrade Mezzio / Doctrine, and use composer2 too

File: composer.json
Comment (v-noskov) [inline composer.json:27]: I think it should be upgraded to satisfy composer requirements
```suggestion
        "doctrine/coding-standard": "^8.1",
```

**Comment by internalsystemerror** [comment]
If this PR is merged, it will also close https://github.com/Roave/psr-container-doctrine/issues/37

**Comment by thexpand** [comment]
What is the status of this PR, is it going to be merged?

File: composer.json
Comment (PowerKiKi) [reply composer.json:27]: I can confirm, after testing locally, that this would solve the failing Travis build.

@edigu could you please accept this suggested change ?

**Comment by PowerKiKi** [comment]
@edigu if you are able to merge https://github.com/edigu/psr-container-doctrine/pull/1, then this PR will pass Travis checks and might have more chance to get merged.

**Comment by edigu** [comment]
> @edigu if you are able to merge [edigu#1](https://github.com/edigu/psr-container-doctrine/pull/1), then this PR will pass Travis checks and might have more chance to get merged.

Hey @PowerKiKi thanks for your help! I have merged your changes into my fork. Besides analysis failures, I have nothing more to add or change at the moment. Perhaps it's time to wait for builds and get a second review then a green light from Roave members.

**Comment by PowerKiKi** [comment]
@asgrim, @Ocramius could you please have a look at this PR ?

Travis is failing as expected, because:

- this PR contains breaking change, so roave/backward-compatibility-check [correctly fails](https://travis-ci.org/github/Roave/psr-container-doctrine/jobs/765989729#L602-L610)
- roave/backward-compatibility-check correctly [cannot be run on PHP 8.0](https://travis-ci.org/github/Roave/psr-container-doctrine/jobs/765989730#L458-L462)

The second issue is solved in the [migration to GitHub Actions](https://github.com/Roave/psr-container-doctrine/pull/38), as you can see [in my fork](https://github.com/PowerKiKi/psr-container-doctrine/actions/runs/710884964). I did not include that migration in this PR to avoid scope creep. Let me know if you prefer to have everything in a single PR.

IMHO this can be merged, and then immediately be followed by https://github.com/Roave/psr-container-doctrine/pull/38.

**Review by asgrim — APPROVED** [review]

**Comment by asgrim** [comment]
Thank you everyone for your work on this :)

**Pull request merged**
